src/Microsoft.AspNet.OData.Shared/Query/Expressions/SelectExpandWrapperJsonConverter.cs

```diff
@@ -0,0 +1,308 @@
+        (IEdmModel model, IEdmStructuredType type) => new JsonPropertyNameMapper(model, type);
+
+        /// <inheritdoc/>
+        public override bool CanConvert(Type typeToConvert)
+        {
+            if (typeToConvert == null)
+            {
+                throw Error.ArgumentNull("objectType");
+            }
+
+            if (!typeToConvert.IsGenericType)
+            {
+                return false;
+            }
+
+            return typeof(ISelectExpandWrapper).IsAssignableFrom(typeToConvert);
+        }
+
+        /// <inheritdoc/>
+        public override JsonConverter CreateConverter(
+            Type typeToConvert,
+            JsonSerializerOptions options)
+        {
+            if (typeToConvert == null || !typeToConvert.IsGenericType)
+            {
+                return null;
+            }
+
+            Type genericType = typeToConvert.GetGenericTypeDefinition();
+            Type type = typeToConvert.GetGenericArguments()[0];
+
+            if (genericType == typeof(SelectExpandBinder.SelectSome<>))
+            {
+                return (JsonConverter)Activator.CreateInstance(
+                    typeof(SelectSomeOfTJsonConverter<>).MakeGenericType(new Type[] { type }));
+            }
+            else if (genericType == typeof(SelectExpandBinder.SelectSomeAndInheritance<>))
+            {
+                return (JsonConverter)Activator.CreateInstance(
+                    typeof(SelectSomeAndInheritanceOfTJsonConverter<>).MakeGenericType(new Type[] { type }));
+            }
+            else if (genericType == typeof(SelectExpandBinder.SelectAll<>))
+            {
+                return (JsonConverter)Activator.CreateInstance(
+                    typeof(SelectAllOfTJsonConverter<>).MakeGenericType(new Type[] { type }));
+            }
+            else if (genericType == typeof(SelectExpandBinder.SelectAllAndExpand<>))
+            {
+                return (JsonConverter)Activator.CreateInstance(
+                    typeof(SelectAllAndExpandOfTJsonConverter<>).MakeGenericType(new Type[] { type }));
+            }
+            else if (genericType == typeof(SelectExpandWrapper<>))
+            {
+                return (JsonConverter)Activator.CreateInstance(
+                    typeof(SelectExpandWrapperOfTJsonConverter<>).MakeGenericType(new Type[] { type }));
+            }
+
+            return null;
+        }
+
+        private class JsonPropertyNameMapper : IPropertyMapper
+        {
+            private IEdmModel _model;
+            private IEdmStructuredType _type;
+
+            public JsonPropertyNameMapper(IEdmModel model, IEdmStructuredType type)
+            {
+                _model = model;
+                _type = type;
+            }
+
+            public string MapProperty(string propertyName)
+            {
+                IEdmProperty property = _type.Properties().Single(s => s.Name == propertyName);
```

Comment:
Do we need to validate that the properties names are unique at this point? Shouldn't this be handled by EdmLib or the writer? Why not use `.First()` or `.FirstOrDefault()`?